packages/compiler/src/testing/expect.ts

```diff
@@ -56,86 +56,224 @@ export interface DiagnosticMatch {
 /**
  * Validate the diagnostic array contains exactly the given diagnostics.
  * @param diagnostics Array of the diagnostics
+ * @param match The diagnostic match(es) to validate against
+ * @param options Options for validation
+ * @param options.strict When true, expects exactly the same number of diagnostics as matches
+ * @param options.fixedOrder When true, diagnostics must appear in the same order as matches. When false, order is ignored.
  */
 export function expectDiagnostics(
   diagnostics: readonly Diagnostic[],
   match: DiagnosticMatch | DiagnosticMatch[],
-  options = {
-    strict: true,
-  },
+  options: {
+    strict?: boolean;
+    fixedOrder?: boolean;
+  } = {},
 ) {
+  // Set defaults
+  const { strict = true, fixedOrder = true } = options;
+
   const array = isArray(match) ? match : [match];
 
-  if (options.strict && array.length !== diagnostics.length) {
+  if (strict && array.length !== diagnostics.length) {
     fail(
       `Expected ${array.length} diagnostics but found ${diagnostics.length}:\n ${formatDiagnostics(
         diagnostics,
       )}`,
     );
   }
-  for (let i = 0; i < array.length; i++) {
-    const diagnostic = diagnostics[i];
-    const expectation = array[i];
-    const sep = "-".repeat(100);
-    const message = `Diagnostics found:\n${sep}\n${formatDiagnostics(diagnostics)}\n${sep}`;
-    if (expectation.code !== undefined) {
-      strictEqual(
-        diagnostic.code,
-        expectation.code,
-        `Diagnostic at index ${i} has non matching code.\n${message}`,
+
+  if (fixedOrder) {
```

Comment:
The approach we decided here was to sort the diagnostics so they diff can still be used. 
The reason I didn't want to add this at it was was exactly because it made the test experience worse when it didn't find the match and in general it was bad pattern to have multiple diagnostics tested at the same time. We decided that sorting the diagnositcs and expectation would still provide the benefits while not loosing the diff